### PR TITLE
Decouple JsonTypeConverter from TypeConverter

### PR DESCRIPTION
**Pull request by North101**
WIP

### CODE REVIEW

**Comment by North101** [comment]
Ok I think I have this working now. I tried to keep it backwards compatible but I wasn't able to without doing a lot more changes I wasn't comfortable.

**Comment by North101** [comment]
Nevermind, found a few issues

File: drift/lib/src/runtime/types/converters.dart [hunk 1]
```diff
@@ -35,25 +35,30 @@ abstract class TypeConverter<D, S> {
 /// `.drift` files) refers to a type converter that inherits from
 /// [JsonTypeConverter], it will also be used for the conversion from and to
 /// JSON.
-mixin JsonTypeConverter<D, S> on TypeConverter<D, S> {
+abstract class JsonTypeConverter<D, S, J> extends TypeConverter<D, S> {
```
Comment (simolus3) [inline drift/lib/src/runtime/types/converters.dart]: One way to make this non-breaking might be to introduce this as a new superclass (imagine `JsonTypeConverterWithDifferentTypes` but with a less cumbersome name).

Them, `typealias JsonTypeConverter<D, S> = JsonTypeConverterWithDifferentTypes<D, S, S>;` would make this non-breaking I believe? Or just `class JsonTypeConverter<D, S> extends JsonTypeConverterWithDifferentTypes<D, S, S>`. I might still be overlooking something, but maybe that works to make this non-breaking.

**Review by simolus3 — COMMENTED** [review]
I like the idea and the implementation in the generator.

I have an idea to make this non-breaking (but I didn't fully think this through, might not be entirely sound). Still, I think it could be worth trying out.

File: drift/lib/src/runtime/types/converters.dart (hunk 1 above)
Comment (North101) [reply drift/lib/src/runtime/types/converters.dart]: I'll give this a go later

Comment (North101) [reply drift/lib/src/runtime/types/converters.dart]: I think it's backwards compatible now. Thanks for the suggestion!

Not sure what to name `JsonTypeConverterWithDifferentTypes`, it seems a little verbose.

File: drift/lib/src/runtime/types/converters.dart [hunk 2]
```diff
@@ -33,17 +33,47 @@ abstract class TypeConverter<D, S> {
 /// to SQL (and vice-versa).
 /// When a [BuildGeneralColumn.map] column (or a `MAPPED BY` constraint in
 /// `.drift` files) refers to a type converter that inherits from
-/// [JsonTypeConverter], it will also be used for the conversion from and to
+/// [JsonTypeConverterWithDifferentTypes], it will also be used for the conversion from and to
 /// JSON.
-mixin JsonTypeConverter<D, S> on TypeConverter<D, S> {
+mixin JsonTypeConverterWithDifferentTypes<D, S, J> on TypeConverter<D, S> {
```
Comment (simolus3) [inline drift/lib/src/runtime/types/converters.dart]: Hard to think of a better name here... Maybe `BaseJsonTypeConverter` could work?

File: drift/lib/src/runtime/types/converters.dart (hunk 1 above)
Comment (simolus3) [reply drift/lib/src/runtime/types/converters.dart]: Nice!

File: drift/lib/src/runtime/types/converters.dart (hunk 2 above)
Comment (North101) [reply drift/lib/src/runtime/types/converters.dart]: Maybe just `JsonTypeConverter2`?

Comment (simolus3) [reply drift/lib/src/runtime/types/converters.dart]: Sounds good 👍

**Review by simolus3 — APPROVED** [review]
Thanks for your contribution!

**Comment by North101** [comment]
> Thanks for your contribution!

No problem. Thanks!

Maybe in the next breaking release we could rename:
`JsonTypeConverter2` to `JsonTypeConverter`
`JsonTypeConverter` to `SimpleJsonTypeConverter`

**Pull request merged**